qutip/core/qobj.py

```diff
@@ -296,58 +271,110 @@ class Qobj:
         elif isinstance(arg, Qobj):
-            self.dims = dims or arg.dims.copy()
             self._data = arg.data.copy() if copy else arg.data
-        elif arg is None or isinstance(arg, numbers.Number):
-            self.dims = dims or [[1], [1]]
-            size = np.prod(self.dims[0])
-            if arg is None:
-                self._data = _data.zeros(size, size)
-            else:
-                self._data = _data.identity(size, scale=complex(arg))
+            self._dims = dims or arg._dims
+            if arg._isherm is not None:
+                self._isherm = arg._isherm
+            if arg._isunitary is not None:
+                self._isunitary = arg._isunitary
         else:
             self._data = _data.create(arg, copy=copy)
+            if (
+                dims
+                and self._data.shape != dims.shape
+                and self._data.shape == dims.shape[::-1]
+            ):
+                self._data = _data.transpose(self._data)
             self.dims = dims or [[self._data.shape[0]], [self._data.shape[1]]]
 
     def __init__(self, arg=None, dims=None, type=None,
                  copy=True, superrep=None, isherm=None, isunitary=None):
-        self._initialize_data(arg, dims, copy)
-        self.type = type or type_from_dims(self.dims)
         self._isherm = isherm
         self._isunitary = isunitary
+        self._superrep = None
+        if isinstance(dims, list):
+            dims = Dimensions(dims)
+        self._initialize_data(arg, dims, copy)
 
-        if self.type == 'super' and type_from_dims(self.dims) == 'oper':
-            if self._data.shape[0] != self._data.shape[1]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare data of shape ",
-                    repr(self._data.shape),
-                ]))
-            root = int(np.sqrt(self._data.shape[0]))
-            if root * root != self._data.shape[0]:
-                raise ValueError("".join([
-                    "cannot build superoperator from nonsquare subspaces ",
-                    "of size ",
-                    repr(self._data.shape[0]),
-                ]))
-            self.dims = [[[root]]*2]*2
-        if self.type in ['super', 'operator-ket', 'operator-bra']:
-            superrep = superrep or 'super'
+        # Dims are guessed from the data and need to be changed to super.
+        if (
+            type in ['super', 'operator-ket', 'operator-bra']
+            and self._dims.type in ['oper', 'ket', 'bra']
+        ):
+            root_right = int(np.sqrt(self._data.shape[0]))
+            root_left = int(np.sqrt(self._data.shape[1]))
+            if (
+                root_right * root_right != self._data.shape[0]
+                and root_left * root_left != self._data.shape[1]
+            ):
+                raise ValueError(
+                    "cannot build superoperator from nonsquare subspaces"
```

Comment:
Not too sure how to properly explain it.
`spre(NxM) @ spost(MxN)` is fine but `spre(NxN) @ spost(MxM)` is not.
Both the input and output state of the super operator must be square matrix.

Reply:
If dims where not passed, dims where created to match the shape of the data.
If dims where passed, the check that their shape match the data shape was done in _initialize_data.

Reply:
Hmm. The maybe the issue is that the check is sort of split between here and `_initialize_data`. Can we somehow make it so that the check happens in one place, or that it's repeated more fully here?

Reply:
It looks like the relevant code is gone, but I don't know where the check went. How did you decide to resolve this?

Reply:
I tried removing the type parameter and setter.
This mean that 1x1 are stuck at "scalar" without ways to change it and when creating super operators, setting dims properly is needed but were quite few places where it was not the case.